c10/util/complex.h

```diff
@@ -0,0 +1,477 @@
+#pragma once
+
+#include <complex>
+#include <iostream>
+
+#include <c10/macros/Macros.h>
+
+#if defined(__CUDACC__) || defined(__HIPCC__)
+#include <thrust/complex.h>
+#endif
+
+namespace c10 {
+
+// c10::complex is an implementation of complex numbers that aims
+// to work on all devices supported by PyTorch
```

Comment:
We should test that this `is_standard_layout` when instantiated with float/double. It probably isn't POD, due to the copy assignment?

Reply:
I added the test, and I also added a test testing that `std::complex` can be `reinterpret_cast`ed to `c10::complex`